doc/man6/vgatest.6
.TH vgatest 6 "01 January 2001" "Svgalib (>= 1.9.13)" "Svgalib User Manual"
.SH NAME
vgatest \- makes basic tests on any svgalib graphics mode
.SH SYNOPSIS

.B vgatest

.SH DESCRIPTION
Mode test program.  First the program gives a list of the supported
modes, then you enter a number and 
.B vgatest 
displays a test pattern in that mode.

The test pattern consists of a sequence of crosses (in different colors,
if the mode support has different colors) in the top left corner.

Below you see either four horizontal color bars (white, red, green,
and blue with increasing intensity from left to right) or vertical
stripes.

If the mode has 32K or more colors, six squares will appear on top of
that, each showing some different smooth color shades

The whole pattern is enclosed in a white border around the edge of the
screen.

See below for details of the test pattern in case you need to verify that
it is being displayed correctly or diagnose a problem with the display.

It also shows some details from
.BR vga_getmodeinfo (3).

To terminate the display and exit the program, hit any key.  If that key
is 'd', 
.B vgatest
dumps the values of all the SVGA registers to Standard Output, in the
manner of vga_dumpregs() before exiting.

This demo is part of svgalib and can be found in the
.I demos/
subdirectory of the original svgalib distribution. However, it is not installed in the system
by default, s.t. it is unclear where you can find it if your svgalib was installed by some
linux distribution. Even then, when you have the demo on your system, you probably won't have
the sources s.t. it is only of limited use for you.

In case of any such problem, simply get an svgalib distribution from the net. You even
don't need to install it. Just
.B make
in the
.I demos/
subdirecty. As of this writing,
.I svgalib-1.2.12.tar.gz
is the latest version and can be retrieved by ftp from
.IR "sunsite.unc.edu" " at " "/pub/Linux/libs/graphics"
and
.IR "tsx-11.mit.edu" " at " "/pub/linux/sources/libs"
which will most probably be mirrored by a site close to you.

.SS The Test Pattern
Here are the details of the test pattern that 
.B vgatest
displays.

The screen is surrounded by a white frame which is at the very edges of 
the screen and one pixel thick.

At the top of the screen is a set of 16 overlapping crosses, lined up
horizontally 5 pixels apart.  Each cross is composed of a one pixel
thick line sloping down at 45 degrees and one sloping up at 45 degrees
meeting in the center.  The top of each of these lines is at the 11th
raster line on the screen and the bottom is at the 90th, so the
crosses are 80 lines high.  The leftmost pixel of the crosses is in
the 11th column of the screen and the rightmost is in the 165th
column.  The lines are in multiple colors, but each line is one color
(except where it intersects another line).

Below the crosses are either 4 horizontal bands of color or vertical
bars.  If the mode has 2 or 16 colors, you get the vertical bars.  If
the mode has 256, 32K, 64K, or 16M colors, you get the horizontal
bands of color.

For a 2 color mode, the vertical bars are are one pixel wide white
bars, spaced 4 pixels apart all the way across the screen on a black
background.  The first white bar is in the 3rd column.  (But don't
forget the white border described above).

For a 16 color mode, the vertical bars are one pixel wide and
contiguous, filling the 3rd column from the left through the 3rd
column from the right.  The bars cycle through each of the 16 colors
from left to right.

The vertical bars start in the 101st raster line and end in the 3rd 
line from the bottom of the screen.

For a higher color mode, the color bands fill the 3rd column from the
left through the 3rd column from the right.  (leaving a column of
black and the aforementioned white border at the edges).  The bands
are all the same height with nothing between them.  They are as large
as will fit starting in the 101st line of the screen and ending at or
before the 3rd line from the bottom.  So depending on the number of
lines on the screen, there are 1, 2, 3, or 4 black lines between the
lowest bar and the white bottom border.

The hues of the bands are, from top to bottom, white, red,
green, and blue.  Each band goes from zero to maximal brightness from
left to right.

In addition to the color bands, for a mode with 32K or more colors,
six squares, 64 pixels on a side are arranged in a matrix centered on
the screen, spaced 32 pixels apart.  They replace any other part of
the test pattern.  These are actual squares only if your monitor
displays square pixels in this mode, i.e. 64 columns is the same
length as 64 lines.  Normal monitors, properly adjusted, display
square pixels for 1024 x 768 modes, but may not for other geometries.

Each square contains 4096 different colors, one pixel per color, in
smooth transition.  In each top square, one color component (red,
green, or blue) is zero another varies linearly in the vertical
direction and the other varies linearly in the Y direction.  The
bottom squares are the same except that one color component is maximum
instead of zero.

.SH SEE ALSO

.BR svgalib (7),
.BR vgagl (7),
.BR libvga.config (5),
.BR vga_getmodeinfo (3),
.BR threed (6),
.BR accel (6),
.BR bg_test (6),
.BR eventtest (6),
.BR forktest (6),
.BR fun (6),
.BR keytest (6),
.BR mousetest (6),
.BR scrolltest (6),
.BR speedtest (6),
.BR spin (6),
.BR testaccel (6),
.BR testgl (6),
.BR testlinear (6),
.BR plane (6),
.BR wrapdemo (6)

.SH AUTHOR

This manual page was edited by Michael Weller <eowmob@exp-math.uni-essen.de>. The
exact source of the referenced demo as well as of the original documentation is
unknown.

It is very likely that both are at least to some extent are due to
Harm Hanemaayer <H.Hanemaayer@inter.nl.net>.

Occasionally this might be wrong. I hereby
asked to be excused by the original author and will happily accept any additions or corrections
to this first version of the svgalib manual.


demos/vgatest.c
/* From VGAlib, changed for svgalib */
/* partially copyrighted (C) 1993 by Hartmut Schirmer */


#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>		/* for usleep( long ) */
#include <string.h>
#include "vga.h"

static unsigned char line[2048 * 3];

static void
drawSquares(int const xmax, int const ymax) 
{
    unsigned int const center_x = xmax/2;
        /* The column just to the left of the center (assuming
           even number of columns)
        */
    unsigned int const center_y = ymax/2;
        /* The line just above the center (assuming
           even number of lines)
        */
            
    int x;

    for (x = 0; x < 64; x++) {
        int y;
        for (y = 0; y < 64; y++) {
            /* Top 3 squares */
            vga_setrgbcolor(x * 4 + 3, y * 4 + 3, 0);
            vga_drawpixel(center_x - 32 - 32 - 64 + x, center_y - 80 + y);
            vga_setrgbcolor(x * 4 + 3, 0, y * 4 + 3);
            vga_drawpixel(center_x -  32 + x,          center_y - 80 + y);
            vga_setrgbcolor(0, x * 4 + 3, y * 4 + 3);
            vga_drawpixel(center_x + 32 + 32 + x,      center_y - 80 + y);
            
            /* Bottom 3 squares */
            vga_setrgbcolor(x * 4 + 3, y * 4 + 3, 255);
            vga_drawpixel(center_x - 32 - 32 - 64 + x, center_y + 16 + y);
            vga_setrgbcolor(x * 4 + 3, 255, y * 4 + 3);
            vga_drawpixel(center_x -  32 + x,          center_y + 16 + y);
            vga_setrgbcolor(255, x * 4 + 3, y * 4 + 3);
            vga_drawpixel(center_x + 32 + 32 + x,      center_y + 16 + y);
        }
    }
}



static void testmode(int mode)
{
    int xmax, ymax, i, x, y, yw, ys, c;
    vga_modeinfo *modeinfo;

    vga_setmode(mode);

    modeinfo = vga_getmodeinfo(mode);

    printf("Width: %d  Height: %d  Colors: %d\n",
	   modeinfo->width,
	   modeinfo->height,
	   modeinfo->colors);
    printf("DisplayStartRange: %xh  Maxpixels: %d  Blit: %s\n",
	   modeinfo->startaddressrange,
	   modeinfo->maxpixels,
	   modeinfo->haveblit ? "YES" : "NO");

#ifdef TEST_MODEX
    if (modeinfo->colors == 256)
	printf("Switching to ModeX ... %s\n",
	       (vga_setmodeX()? "done" : "failed"));
#endif

    vga_screenoff();

    xmax = vga_getxdim() - 1;
    ymax = vga_getydim() - 1;

    vga_setcolor(vga_white());
    vga_drawline(0, 0, xmax, 0);
    vga_drawline(xmax, 0, xmax, ymax);
    vga_drawline(xmax, ymax, 0, ymax);
    vga_drawline(0, ymax, 0, 0);

    /* Draw crosses */
    for (i = 0; i <= 15; i++) {
	vga_setegacolor(i);
	vga_drawline(10 + i * 5, 10, 89 + i * 5, 89);
    }
    for (i = 0; i <= 15; i++) {
	vga_setegacolor(i);
	vga_drawline(89 + i * 5, 10, 10 + i * 5, 89);
    }

    vga_screenon();

    ys = 100;
    yw = (ymax - 100) / 4;
    switch (vga_getcolors()) {
    case 256:
        /* Draw horizontal color bands using palette */
	for (i = 0; i < 60; ++i) {
	    c = (i * 64) / 60;
	    vga_setpalette(i + 16, c, c, c);
	    vga_setpalette(i + 16 + 60, c, 0, 0);
	    vga_setpalette(i + 16 + (2 * 60), 0, c, 0);
	    vga_setpalette(i + 16 + (3 * 60), 0, 0, c);
	}
	line[0] = line[xmax] = 15;
	line[1] = line[xmax - 1] = 0;
	for (x = 2; x < xmax - 1; ++x)
	    line[x] = (((x - 2) * 60) / (xmax - 3)) + 16;
	for (y = ys; y < ys + yw; ++y)	/* gray */
	    vga_drawscanline(y, line);
	for (x = 2; x < xmax - 1; ++x)
	    line[x] += 60;
	ys += yw;
	for (y = ys; y < ys + yw; ++y)	/* red */
	    vga_drawscanline(y, line);
	for (x = 2; x < xmax - 1; ++x)
	    line[x] += 60;
	ys += yw;
	for (y = ys; y < ys + yw; ++y)	/* green */
	    vga_drawscanline(y, line);
	for (x = 2; x < xmax - 1; ++x)
	    line[x] += 60;
	ys += yw;
	for (y = ys; y < ys + yw; ++y)	/* blue */
	    vga_drawscanline(y, line);
	break;

    case 1 << 15:
    case 1 << 16:
    case 1 << 24:
        /* Draw horizontal color bands in RGB */
	for (x = 2; x < xmax - 1; ++x) {
	    c = ((x - 2) * 255) / (xmax - 4);
	    y = ys;
	    vga_setrgbcolor(c, c, c);
	    vga_drawline(x, y, x, y + yw - 1);
	    y += yw;
	    vga_setrgbcolor(c, 0, 0);
	    vga_drawline(x, y, x, y + yw - 1);
	    y += yw;
	    vga_setrgbcolor(0, c, 0);
	    vga_drawline(x, y, x, y + yw - 1);
	    y += yw;
	    vga_setrgbcolor(0, 0, c);
	    vga_drawline(x, y, x, y + yw - 1);
	}
        drawSquares(xmax, ymax);
	break;
    default:
        /* Draw vertical color bars */
	if (vga_getcolors() == 16) {
	    for (i = 0; i < xmax - 1; i++)
		line[i] = (i + 2) % 16;
	    line[0] = line[xmax] = 15;
	    line[1] = line[xmax - 1] = 0;
	}
	if (vga_getcolors() == 2) {
	    for (i = 0; i <= xmax; i++)
		line[i] = 0x11;
	    line[0] = 0x91;
	}
	for (i = 100; i < ymax - 1; i++)
	    vga_drawscanline(i, line);
	break;

    }

    if (getchar() == 'd')
	vga_dumpregs();

}

int main(int argc, char *argv[])
{
    int mode, mode2;
    int i, high;

    vga_init();			/* Initialize. */

    mode = -1;
	mode2= -1;
	
	if(argc==2) {
		mode = atoi(argv[1]);
	} else if(argc==3) {
		mode = atoi(argv[1]);
		mode2= atoi(argv[2]);
	}
		

    if (mode == -1) {
	printf("Choose one of the following video modes: \n");

	high = 0;
	for (i = 1; i <= vga_lastmodenumber(); i++)
	    if (vga_hasmode(i)) {
		vga_modeinfo *info;
		char expl[100];
		const char *cols = NULL;

		*expl = '\0';
		info = vga_getmodeinfo(i);
		switch (info->colors) {
		case 2:
		    cols = "2";
		    strcpy(expl, "1 bitplane, monochrome");
		    break;
		case 16:
		    cols = "16";
		    strcpy(expl, "4 bitplanes");
		    break;
		case 256:
		    if (i == G320x200x256)
			strcpy(expl, "packed-pixel");
		    else if (i == G320x240x256
			     || i == G320x400x256
			     || i == G360x480x256)
			strcpy(expl, "Mode X");
		    else
			strcpy(expl,
			       "packed-pixel, banked");
		    break;
		case 1 << 15:
		    cols = "32K";
		    strcpy(expl, "5-5-5 RGB, blue at LSB, banked");
		    break;
		case 1 << 16:
		    cols = "64K";
		    strcpy(expl, "5-6-5 RGB, blue at LSB, banked");
		    break;
		case 1 << 24:
		    cols = "16M";
		    if (info->bytesperpixel == 3) {
			if (info->flags & RGB_MISORDERED)
			    strcpy(expl, "8-8-8 BGR, red byte first, banked");
			else
			    strcpy(expl, "8-8-8 RGB, blue byte first, banked");
		    } else if (info->flags & RGB_MISORDERED)
			strcpy(expl, "8-8-8 RGBX, 32-bit pixels, X byte first, banked");
		    else
			strcpy(expl, "8-8-8 XRGB, 32-bit pixels, blue byte first, banked");
		    break;
		}
		if (info->flags & IS_INTERLACED) {
		    if (*expl != '\0')
			strcat(expl, ", ");
		    strcat(expl, "interlaced");
		}
		if (info->flags & IS_DYNAMICMODE) {
		    if (*expl != '\0')
			strcat(expl, ", ");
		    strcat(expl, "dynamically loaded");
		}
		high = i;
		printf("%5d: %dx%d, ",
		       i, info->width, info->height);
		if (cols == NULL)
		    printf("%d", info->colors);
		else
		    printf("%s", cols);
		printf(" colors ");
		if (*expl != '\0')
		    printf("(%s)", expl);
		printf("\n");
	    }
	printf("Enter mode number (1-%d): ", high);
	scanf("%d", &mode);
	getchar();
	printf("\n");

	if (mode < 1 || mode > GLASTMODE) {
	    printf("Error: Mode number out of range \n");
	    exit(-1);
	}
    }
    if (vga_hasmode(mode)) {
		testmode(mode);
		if(mode2!=-1 && vga_hasmode(mode2)) {
			testmode(mode2);
		}
	} else {
		printf("Error: Video mode not supported by driver\n");
		exit(-1);
    }
    
	vga_setmode(TEXT);
    
    return 0;
}
